src/seq.rs

```diff
@@ -9,61 +9,337 @@
+    /// apply the indices to the slice.
+    /// 
+    /// Although the elements are selected randomly, the order of returned
+    /// elements is neither stable nor fully random. If random ordering is
+    /// desired, either use `partial_shuffle` or use this method and shuffle
+    /// the result. If stable order is desired, use `sample_indices`, sort the
+    /// result, then apply to the slice.
+    /// 
+    /// Complexity is expected to be the same as `sample_indices`.
+    #[cfg(feature = "alloc")]
+    fn choose_multiple<R>(&self, rng: &mut R, amount: usize) -> SliceChooseIter<Self, Self::Item>
+        where R: Rng + ?Sized;
+
+    /// Shuffle a mutable slice in place.
+    /// 
+    /// Depending on the implementation, complexity is expected to be `O(1)`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use rand::thread_rng;
+    /// use rand::seq::SliceExt;
+    ///
+    /// let mut rng = thread_rng();
+    /// let mut y = [1, 2, 3, 4, 5];
+    /// println!("Unshuffled: {:?}", y);
+    /// y.shuffle(&mut rng);
+    /// println!("Shuffled:   {:?}", y);
+    /// ```
+    fn shuffle<R>(&mut self, rng: &mut R) where R: Rng + ?Sized;
+
+    /// Shuffle a slice in place, but exit early.
+    ///
+    /// Returns two mutable slices from the source slice. The first contains
+    /// `amount` elements randomly permuted. The second has the remaining
+    /// elements that are not fully shuffled.
+    ///
+    /// This is an efficient method to select `amount` elements at random from
+    /// the slice, provided the slice may be mutated.
+    ///
+    /// If you only need to chose elements randomly and `amount > self.len()/2`
+    /// then you may improve performance by taking
+    /// `amount = values.len() - amount` and using only the second slice.
+    ///
+    /// If `amount` is greater than the number of elements in the slice, this
+    /// will perform a full shuffle.
+    ///
+    /// Depending on the implementation, complexity is expected to be `O(m)`,
+    /// where `m = amount`.
+    fn partial_shuffle<R>(&mut self, rng: &mut R, amount: usize)
+        -> (&mut [Self::Item], &mut [Self::Item]) where R: Rng + ?Sized;
+}
+
+/// Extension trait on iterators, providing random sampling methods.
+pub trait IteratorExt: Iterator + Sized {
+    /// Choose one element at random from the iterator.
+    ///
+    /// Returns `None` if and only if the iterator is empty.
+    /// 
+    /// Complexity is `O(n)`, where `n` is the length of the iterator.
+    /// This likely consumes multiple random numbers, but the exact number
+    /// is unspecified.
+    fn choose<R>(mut self, rng: &mut R) -> Option<Self::Item>
+        where R: Rng + ?Sized
+    {
+        if let Some(elem) = self.next() {
+            let mut result = elem;
+            
+            // Continue until the iterator is exhausted
+            for (i, elem) in self.enumerate() {
+                let k = rng.gen_range(0, i + 2);
+                if k == 0 {
```

Comment:
`gen_ratio` gets a big speed-up:
```
# current
test seq_iter_choose_from_100                ... bench:       1,053 ns/iter (+/- 55)
# using (i + 2) as u32
test seq_iter_choose_from_100                ... bench:         491 ns/iter (+/- 16)
# using TryFrom
test seq_iter_choose_from_100                ... bench:         526 ns/iter (+/- 14)
```
Unfortunately TryFrom is still not stable, and also doesn't support working with `usize` directly (https://github.com/rust-lang/rust/issues/49415). I'm not certain we should use `as` here and because this is an iterator without known exact size we have no choice but to check each time for correctness.

Reply:
Yeah. Let's stick with what you have. This might be an argument for making `gen_ratio` generic over type though. But lets worry about that separately.